Use gzip and snappy compressor pools

kriskowal opened the pull request:
This change anticipates performance problems with compression by pooling readers and writers.

The implementation follows the form but not the letter of the corresponding GZIP compressor from gRPC-go.

https://github.com/grpc/grpc-go/blob/master/encoding/gzip/gzip.go

- Notably, we do not interact with the gRPC global registry.
- The compression level is determined once at time of construction.
- We do not use a closure to populate the pool. We consequently can see coverage for both creation and reuse of pool readers and writers.
- Nonzero tests.

kriskowal commented on compressor/gzip/gzip.go:
```diff
@@ -18,32 +18,197 @@
+func (r *reader) Read(buf []byte) (n int, err error) {
+	n, err = r.reader.Read(buf)
+	if err == io.EOF {
+		r.pool.Put(r)
```
I could allow for premature close. I’d need to write a method to attenuate a Compressor to a GRPCCompressor so it could be registered. gRPC would not use it, but we’d have more liberty in a future version.

prashantv replied:
Yeah, it might be a cleaner approach longer term to add a `Close` method, and only put the reader back in the pool on `Close()`. If we do that, in gRPC, we'd need the wrapper to do the `if err == io.EOF { Close(..) }` though

aignas replied:
Nice to see that I was not the only one who was confused about gRPC not using the Close method. SGTM.

kriskowal commented:
I’ve changed the YARPC interface to have a ReadCloser and the convention that the reader must be Closed. I’ve added a gRPC adapter that implicitly calls Close on the underlying reader when gRPC calls Read and receives EOF.

kriskowal commented:
I added the wrapper for gRPC. It’s unfortunate but the waste is bearable, knowing that gRPC-go is not dictating our design.

aignas commented on api/transport/compression.go:
```diff
@@ -30,5 +30,5 @@ import "io"
 	Name() string
 	Compress(w io.Writer) (io.WriteCloser, error)
-	Decompress(r io.Reader) (io.Reader, error)
+	Decompress(r io.Reader) (io.ReadCloser, error)
```
This is great, thanks for having this.
